server/app/services/applicant/ApplicantServiceImpl.java

```diff
@@ -358,6 +367,138 @@ public ImmutableList<Application> getAllApplications() {
+            .collect(ImmutableList.toImmutableList());
+
+    return applicationsFuture.thenApplyAsync(
+        applications -> {
+          checkForDuplicateDrafts(applications);
+          return relevantProgramsForApplicant(activePrograms, applications);
+        },
+        httpExecutionContext.current());
+  }
+
+  private RelevantPrograms relevantProgramsForApplicant(
+      ImmutableList<ProgramDefinition> activePrograms, ImmutableSet<Application> applications) {
+    ImmutableMap<String, ProgramDefinition> activeProgramNames =
+        ImmutableMap.copyOf(
+            activePrograms.stream()
+                .collect(Collectors.toMap(ProgramDefinition::adminName, pdef -> pdef)));
+
+    // When new versions of Programs are created, they have distinct IDs but retain the
+    // same adminName. In order to find the most recent draft / active application,
+    // we first group by the unique name rather than the ID.
+    Map<String, Map<LifecycleStage, Optional<Application>>> mostRecentApplicationsByProgram =
+        applications.stream()
+            .collect(
+                Collectors.groupingBy(
+                    a -> {
+                      return a.getProgram().getProgramDefinition().adminName();
+                    },
+                    Collectors.groupingBy(
+                        Application::getLifecycleStage,
+                        // In practice, we don't expect an applicant to have multiple applications
+                        // for a given program / LifecycleStage combination. Grabbing the latest
+                        // application here guards against that case, should it occur.
+                        Collectors.maxBy(Comparator.comparingLong(a -> a.id)))));
+
+    ImmutableList.Builder<ApplicantProgramData> inProgressPrograms = ImmutableList.builder();
+    ImmutableList.Builder<ApplicantProgramData> submittedPrograms = ImmutableList.builder();
+    ImmutableList.Builder<ApplicantProgramData> unappliedPrograms = ImmutableList.builder();
+
+    mostRecentApplicationsByProgram.forEach(
+        (programName, appByStage) -> {
+          Optional<Application> maybeDraftApp =
+              appByStage.getOrDefault(LifecycleStage.DRAFT, Optional.empty());
+          Optional<Application> maybeSubmittedApp =
+              appByStage.getOrDefault(LifecycleStage.ACTIVE, Optional.empty());
+          Optional<Instant> maybeSubmitTime = maybeSubmittedApp.map(Application::getSubmitTime);
+          if (maybeDraftApp.isPresent()) {
+            // We want to ensure that any generated links points to the version
+            // of the program associated with the draft, not the most recent version.
+            // As such, we use the program definition associated with the application.
+            inProgressPrograms.add(
+                ApplicantProgramData.create(
+                    maybeDraftApp.get().getProgram().getProgramDefinition(), maybeSubmitTime));
```

Comment:
You're using the submitted apps time not the draft's time.

Reply:
That's correct. We still want to show when the program was submitted, even if there's a draft version. Additionally, the draft version doesn't have a "last edited" timestamp, as I understand.

Reply:
> Will the UI will distinguish a draft with a previously submitted application?

Yes - it will be in the "In progress" section and contain the title "Submitted 2/2/2022".

> I'm not quite understanding why it makes sense to have the wrong latestApplicationSubmitTime associated with an application. Would it be equally correct to name the field latestSubmittedApplicationTime? If so that is clearer naming imo.

To be clear, the information displayed in the UI is the latest *submitted* application time, so attaching the submitted time for a distinct application to a draft application is still correct with respect to what we want to display in the UI.

That said, I agree that the naming is confusing. Happy to rename.